DigitalLearningSolutions.Data.Tests/Services/LearningHubResourceServiceTests.cs

```diff
@@ -0,0 +1,210 @@
+                A.CallTo(
+                        () => learningHubApiClient.GetBulkResourcesByReferenceIds(
+                            A<IEnumerable<int>>.That.IsSameSequenceAs(resourceReferenceIdsToRetrieve)
+                        )
+                    )
+                    .MustHaveHappenedOnceExactly();
+                A.CallTo(
+                        () => learningResourceReferenceDataService.GetResourceReferenceDetailsByReferenceIds(
+                            A<IEnumerable<int>>._
+                        )
+                    )
+                    .MustNotHaveHappened();
+            }
+        }
+
+        [Test]
+        public async Task GetBulkResourcesByReferenceIds_returns_fallback_data_if_API_throws_exception()
+        {
+            // Given
+            var resource1 = GenerateGenericResource(1);
+            var resource2 = GenerateGenericResource(2);
+            var resourceReferenceIdsToRetrieve = new[] { 1, 2, 3 };
+            var bulkResources = new BulkResourceReferences
+            {
+                ResourceReferences = new List<ResourceReferenceWithResourceDetails>
+                {
+                    resource1,
+                    resource2,
+                },
+                UnmatchedResourceReferenceIds = new List<int> { 3 },
+            };
+
+            A.CallTo(() => learningHubApiClient.GetBulkResourcesByReferenceIds(A<IEnumerable<int>>._))
+                .Throws<Exception>();
+            A.CallTo(
+                () => learningResourceReferenceDataService.GetResourceReferenceDetailsByReferenceIds(
+                    A<IEnumerable<int>>._
+                )
+            ).Returns(new List<ResourceReferenceWithResourceDetails> { resource1, resource2 });
+
+            // When
+            var result =
+                (await learningHubResourceService.GetBulkResourcesByReferenceIds(resourceReferenceIdsToRetrieve))!;
+
+            // Then
+            using (new AssertionScope())
+            {
+                result.BulkResourceReferences.Should().BeEquivalentTo(bulkResources);
+                result.SourcedFromFallbackData.Should().BeTrue();
+                A.CallTo(
+                        () => learningHubApiClient.GetBulkResourcesByReferenceIds(
+                            A<IEnumerable<int>>.That.IsSameSequenceAs(resourceReferenceIdsToRetrieve)
+                        )
+                    )
+                    .MustHaveHappenedOnceExactly();
+                A.CallTo(
+                        () => learningResourceReferenceDataService.GetResourceReferenceDetailsByReferenceIds(
+                            A<IEnumerable<int>>.That.IsSameSequenceAs(resourceReferenceIdsToRetrieve)
+                        )
+                    )
+                    .MustHaveHappenedOnceExactly();
+            }
+        }
+
+        private ResourceReferenceWithResourceDetails GenerateGenericResource(int resourceReferenceId)

```

Comment:
Any reason not to use NBuilder instead?

Reply:
Ah, well I meant instead of having this method. If you just invoke `Builder<ResourceReferenceWithResourceDetails>.CreateNew.Build()` I'd have thought that would suffice in most cases, rather than having this method which specifies all the properties.